services/service-helpers/validator3.js

```diff
@@ -0,0 +1,173 @@
+'use strict';
+/**
+ * Validators are functions which assert certain type.
+ * They can return a string which is generally a simple
+ * type which can then be used to display a helpful error message.
+ * They can also return a function for a more complicated
+ * error message.
+ */
+var isPlainObject = require('is-plain-obj');
+
+var v = {};
+
+function formatMessage(apiName, message) {
+  apiName = apiName ? apiName + ';' : '';
+  return apiName + 'Validation failed! ' + message;
+}
+
+function objectEntries(obj) {
+  return Object.keys(obj || {}).map(function(key) {
+    return { key: key, val: obj[key] };
+  });
+}
+
+v.warn = function(rootValidator, apiName) {
+  return function(value) {
+    var messages = validate(rootValidator, value);
+    // all good
+    if (!messages) {
+      return;
+    }
+
+    var len = messages.length;
+    // this where the validator assertion is
+    var lastElement = messages[len - 1];
+    var path = messages.slice(0, len - 1);
+
+    if (path.length == 0) {
+      // Calling it value since there is no identifiable path
+      path = ['value'];
+    }
+
+    var errorMessage;
+
+    if (typeof lastElement == 'function') {
+      errorMessage = lastElement(apiName, path);
+    } else {
+      var startWith = '';
+      // an array item is invalid
+      if (
+        typeof path[path.length - 1] == 'number' ||
+        typeof path[0] == 'number'
+      ) {
+        startWith = 'item at position ';
+      }
+
+      errorMessage = formatMessage(
+        apiName,
+        startWith + path.join('.') + ' must be ' + lastElement
+      );
+    }
```

Comment:
I think it would clarify the code to extract this if-else block into a separate function that returns `errorMessage`. You could use an early return and reduce some nesting.